extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Because AdSense and DoubleClick are both operated by Google and their A4A
+// implementations share some behavior in common, part of the logic for this
+// implementation is located in the ads/google/a4a directory rather than here.
+// Most other ad networks will want to put their A4A code entirely in the
+// extensions/amp-ad-network-${NETWORK_NAME}-impl directory.
+
+import {AmpA4A} from '../../amp-a4a/0.1/amp-a4a';
+import {
+  extractGoogleAdCreativeAndSignature,
+  getGoogleAdSlotCounter,
+  googleAdUrl,
+  isGoogleAdsA4AValidEnvironment,
+} from '../../../ads/google/a4a/utils';
+import {timer} from '../../../src/timer';
+import {viewportFor} from '../../../src/viewport';
+
+/** @const {string} */
+const DOUBLECLICK_BASE_URL =
+    'https://securepubads.g.doubleclick.net/gampad/ads';
+
+export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
+
+  /**
+   * @param {!Element} element
+   */
+  constructor(element) {
+    super(element);
+  }
+
+  /** @override */
+  isValidElement() {
+    return isGoogleAdsA4AValidEnvironment(this.getWin()) && this.isInAmpAdTag();
```

Comment:
Our a4a implementation doesn't support publishers' use of `remote.html` (see https://github.com/ampproject/amphtml/blob/master/extensions/amp-ad/amp-ad.md#running-ads-from-a-custom-domain).  So isValidElement should return false for any page that contains the `<meta name="amp-3p-iframe-src"...>` tag.